File: examples/java/src/main/java/org/wickra/examples/StrategyRsiMeanReversion.java
package org.wickra.examples;

import org.wickra.Rsi;
import org.wickra.examples.MarketData.Bar;

import java.util.ArrayList;
import java.util.List;

/** Mean reversion: go long when RSI(14) drops below 30, exit when it recovers above 50. */
public final class StrategyRsiMeanReversion {
    public static void main(String[] args) throws Exception {
        Bar[] bars = args.length > 0 ? MarketData.loadOhlcvCsv(args[0]) : MarketData.syntheticCandles(2000);

        try (Rsi rsi = new Rsi(14)) {
            List<Double> returns = new ArrayList<>();
            int trades = 0;
            boolean inPosition = false;
            double entry = 0.0;

            for (Bar b : bars) {
                double value = rsi.update(b.close());
                if (!Double.isFinite(value)) {
                    continue;
                }
                if (!inPosition && value < 30.0) {
                    inPosition = true;
                    entry = b.close();
                    trades++;
                } else if (inPosition && value > 50.0) {
                    returns.add((b.close() - entry) / entry);
                    inPosition = false;
                }
            }

            Equity.print("RSI mean-reversion", Equity.summarize(returns, trades));
        }
    }
}

File: examples/r/strategy_rsi_mean_reversion.R
# Mean reversion: go long when RSI(14) drops below 30, exit when it recovers above 50.
library(wickra)
source("_common.R")

args <- commandArgs(trailingOnly = TRUE)
bars <- if (length(args) >= 1) load_ohlcv_csv(args[1]) else synthetic_candles(2000)

rsi <- Rsi(14)
returns <- numeric(0); trades <- 0L; in_pos <- FALSE; entry <- 0
for (i in seq_len(nrow(bars))) {
  cl <- bars$close[i]
  value <- update(rsi, cl)
  if (!is.finite(value)) next
  if (!in_pos && value < 30) {
    in_pos <- TRUE; entry <- cl; trades <- trades + 1L
  } else if (in_pos && value > 50) {
    returns <- c(returns, (cl - entry) / entry); in_pos <- FALSE
  }
}
print_equity("RSI mean-reversion", summarize_equity(returns, trades))
